From cd678a20fb21b6e5a5b8f05619427dc79aa1246b Mon Sep 17 00:00:00 2001
From: Jerome Soumagne <jerome.soumagne@hpe.com>
Date: Mon, 26 Jan 2026 12:27:27 -0600
Subject: [PATCH 3/3] NA OFI: ensure domain auth key remains valid

---
 src/na/na_ofi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/na/na_ofi.c b/src/na/na_ofi.c
index d385d279..68a443e8 100644
--- a/src/na/na_ofi.c
+++ b/src/na/na_ofi.c
@@ -5220,7 +5220,7 @@ na_ofi_domain_open(const struct na_ofi_fabric *na_ofi_fabric,
             NA_NOMEM, "Could not allocate auth_key");
         memcpy(na_ofi_domain->auth_key, &base_auth_key, sizeof(base_auth_key));
 
-        domain_attr->auth_key = (void *) &base_auth_key;
+        domain_attr->auth_key = (void *) na_ofi_domain->auth_key;
         domain_attr->auth_key_size = auth_key_size;
     }
 
-- 
2.52.0

