diff --git a/src/na/na_ucx.c b/src/na/na_ucx.c
index 96501b2..3c3384e 100644
--- a/src/na/na_ucx.c
+++ b/src/na/na_ucx.c
@@ -3251,16 +3251,9 @@ na_ucx_rma(struct na_ucx_class NA_UNUSED *na_ucx_class, na_context_t *context,
 
     /* There is no need to have a fully resolved address to start an RMA.
      * This is only necessary for two-sided communication. */
-    /* The above assumption is now in question, so the following will resolve
-     * the address if required. */
+    if (!(hg_atomic_get32(&na_ucx_addr->status) & NA_UCX_ADDR_RESOLVED))
+        hg_atomic_or32(&na_ucx_addr->status, NA_UCX_ADDR_RESOLVED);
 
-    /* Check addr to ensure the EP for that addr is still valid */
-    if (!(hg_atomic_get32(&na_ucx_addr->status) & NA_UCX_ADDR_RESOLVED)) {
-        ret = na_ucx_addr_map_update(
-            na_ucx_class, &na_ucx_class->addr_map, na_ucx_addr);
-        NA_CHECK_SUBSYS_NA_ERROR(
-            addr, error, ret, "Could not update NA UCX address");
-    }
     NA_CHECK_SUBSYS_ERROR(rma, na_ucx_addr->ucp_ep == NULL, error, ret,
         NA_ADDRNOTAVAIL, "UCP endpoint is NULL for that address");
 
