diff --git a/CMakeLists.txt b/CMakeLists.txt
index b075819..a764bc6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -229,9 +229,9 @@ function(mercury_set_lib_options libtarget libname libtype var_prefix)
       OUTPUT_NAME_RELEASE        ${LIB_RELEASE_NAME}
       OUTPUT_NAME_MINSIZEREL     ${LIB_RELEASE_NAME}
       OUTPUT_NAME_RELWITHDEBINFO ${LIB_RELEASE_NAME}
-      OUTPUT_NAME_ASAN           ${LIB_DEBUG_NAME}
-      OUTPUT_NAME_TSAN           ${LIB_DEBUG_NAME}
-      OUTPUT_NAME_UBSAN          ${LIB_DEBUG_NAME}
+      OUTPUT_NAME_ASAN           ${LIB_RELEASE_NAME}
+      OUTPUT_NAME_TSAN           ${LIB_RELEASE_NAME}
+      OUTPUT_NAME_UBSAN          ${LIB_RELEASE_NAME}
   )
   if(${libtype} MATCHES "SHARED")
     set_target_properties(${libtarget}
