diff --git a/avocado/core/test.py b/avocado/core/test.py
index 4ff4ee03..73241b4e 100644
--- a/avocado/core/test.py
+++ b/avocado/core/test.py
@@ -461,7 +472,7 @@ class Test(unittest.TestCase):
                          'status', 'time_elapsed',
                          'traceback', 'workdir', 'whiteboard', 'time_start',
                          'time_end', 'running', 'paused', 'paused_msg',
-                         'fail_class', 'params', "timeout"]
+                         'fail_class', 'params', "timeout", 'phase']
         state = {key: getattr(self, key, None) for (key) in preserve_attr}
         state['class_name'] = self.__class__.__name__
         state['job_logdir'] = self.job.logdir
