diff --git a/avocado/core/test.py b/avocado/core/test.py
index 4ff4ee03..73241b4e 100644
--- a/avocado/core/test.py
+++ b/avocado/core/test.py
@@ -62,7 +62,7 @@
                          'status', 'running', 'paused',
                          'time_start', 'time_elapsed', 'time_end',
                          'fail_reason', 'fail_class', 'traceback',
-                         'timeout', 'whiteboard')
+                         'timeout', 'whiteboard', 'phase')
 
 
 class RawFileHandler(logging.FileHandler):
