diff --git a/module/bdev/aio/bdev_aio.c b/module/bdev/aio/bdev_aio.c
index b51d6c83a..e12e32664 100644
--- a/module/bdev/aio/bdev_aio.c
+++ b/module/bdev/aio/bdev_aio.c
@@ -989,6 +989,8 @@ create_aio_bdev(const char *name, const char *filename, uint32_t block_size, boo
 		fdisk->disk.required_alignment = spdk_u32log2(block_size);
 	}
 
+	/* Disable required_alignment forcely */
+	fdisk->disk.required_alignment = 0;
 	if (disk_size % fdisk->disk.blocklen != 0) {
 		SPDK_ERRLOG("Disk size %" PRIu64 " is not a multiple of block size %" PRIu32 "\n",
 			    disk_size, fdisk->disk.blocklen);
