diff --git a/meson_options.txt b/meson_options.txt
index b4608a7..e619e38 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -3,3 +3,7 @@ option('disable-mtab', type : 'boolean', value : false,

 option('udevrulesdir', type : 'string', value : '',
        description: 'Path where udev rules are installed to (Defaults to udevdir specified in udev.pc)')
+
+option('skip-systemfiles', type : 'boolean', value : false,
+       description: 'Whether to install system files or not, incl. udevrules, etc. (Defaults to false)')
+
diff --git a/util/meson.build b/util/meson.build
index ec08c17..3de3056 100644
--- a/util/meson.build
+++ b/util/meson.build
@@ -18,14 +18,14 @@ executable('mount.fuse3', ['mount.fuse.c'],
            install: true,
            install_dir: get_option('sbindir'))

+if not get_option('skip-systemfiles')
+  udevrulesdir = get_option('udevrulesdir')
+  if udevrulesdir == ''
+    udev = dependency('udev')
+    udevrulesdir = join_paths(udev.get_pkgconfig_variable('udevdir'), 'rules.d')
+  endif

-udevrulesdir = get_option('udevrulesdir')
-if udevrulesdir == ''
-  udev = dependency('udev')
-  udevrulesdir = join_paths(udev.get_pkgconfig_variable('udevdir'), 'rules.d')
+  meson.add_install_script('install_helper.sh', get_option('sysconfdir'),
+                           get_option('bindir'), udevrulesdir)
 endif

-meson.add_install_script('install_helper.sh', get_option('sysconfdir'),
-                         get_option('bindir'), udevrulesdir)
-
-

