diff --git a/src/na/na.c b/src/na/na.c
index 08373e3..ea9c9a2 100644
--- a/src/na/na.c
+++ b/src/na/na.c
@@ -121,6 +121,7 @@ na_info_parse(const char *info_string, struct na_info **na_info_ptr)
     /* Initialize NA info */
     na_info->class_name = NULL;
     na_info->protocol_name = NULL;
+    na_info->domain_name = NULL;
     na_info->host_name = NULL;
 
     /* Copy info string and work from that */
@@ -147,10 +148,17 @@ na_info_parse(const char *info_string, struct na_info **na_info_ptr)
         NA_CHECK_ERROR(na_info->class_name == NULL, error, ret, NA_NOMEM_ERROR,
             "Could not duplicate NA info class name");
 
+        token = strtok_r(NULL, NA_CLASS_DELIMITER, &_locator);
+
         /* Get protocol name */
-        na_info->protocol_name = strdup(_locator);
+        na_info->protocol_name = strdup(token);
         NA_CHECK_ERROR(na_info->protocol_name == NULL, error, ret,
             NA_NOMEM_ERROR, "Could not duplicate NA info protocol name");
+
+        /* Get NA domain name */
+        na_info->domain_name = strdup(_locator);
+        NA_CHECK_ERROR(na_info->domain_name == NULL, error, ret,
+            NA_NOMEM_ERROR, "Could not duplicate NA info domain name");
     } else {
         /* Get protocol name */
         na_info->protocol_name = strdup(token);
@@ -195,6 +203,8 @@ na_info_free(struct na_info *na_info)
 
     free(na_info->class_name);
     free(na_info->protocol_name);
+    if (na_info->domain_name)
+        free(na_info->domain_name);
     free(na_info->host_name);
     free(na_info);
 }
@@ -208,6 +218,7 @@ na_info_print(struct na_info *na_info)
 
     printf("Class: %s\n", na_info->class_name);
     printf("Protocol: %s\n", na_info->protocol_name);
+    printf("Domain: %s\n", na_info->domain_name);
     printf("Hostname: %s\n", na_info->host_name);
 }
 #endif
diff --git a/src/na/na.h b/src/na/na.h
index 8d0874e..f26e10a 100644
--- a/src/na/na.h
+++ b/src/na/na.h
@@ -1069,6 +1069,7 @@ NA_Error_to_string(
 struct na_info {
     char *class_name;    /* Class name (e.g., bmi) */
     char *protocol_name; /* Protocol (e.g., tcp, ib) */
+    char *domain_name;
     char *host_name;     /* Host (may be NULL in anonymous mode) */
     /* Additional init info (NULL if no info) */
     const struct na_init_info *na_init_info;
diff --git a/src/na/na_ofi.c b/src/na/na_ofi.c
index 1cd6556..261fe56 100644
--- a/src/na/na_ofi.c
+++ b/src/na/na_ofi.c
@@ -2910,7 +2910,7 @@ na_ofi_initialize(na_class_t *na_class, const struct na_info *na_info,
                 src_addr = na_ofi_sin_addr;
                 src_addrlen = sizeof(*na_ofi_sin_addr);
                 /* Make sure we are using the right domain */
-                strncpy(domain_name, ifa_name, NA_OFI_MAX_URI_LEN - 1);
+                strncpy(domain_name, na_info->domain_name, NA_OFI_MAX_URI_LEN - 1);
                 free(ifa_name);
             } else {
                 /* Allow for passing domain name directly */
