diff --git a/include/spdk/thread.h b/include/spdk/thread.h
index 64b04be..50b473a 100644
--- a/include/spdk/thread.h
+++ b/include/spdk/thread.h
@@ -206,6 +206,13 @@ void spdk_thread_lib_fini(void);
 struct spdk_thread *spdk_thread_create(const char *name, struct spdk_cpuset *cpumask);
 
 /**
+ * Force the current system thread to act as if executing the given SPDK thread.
+ *
+ * \param thread The thread to set.
+ */
+void spdk_set_thread(struct spdk_thread *thread);
+
+/**
  * Release any resources related to the given thread and destroy it. Execution
  * continues on the current system thread after returning.
  *
diff --git a/include/spdk_internal/thread.h b/include/spdk_internal/thread.h
index fe70b48..220d433 100644
--- a/include/spdk_internal/thread.h
+++ b/include/spdk_internal/thread.h
@@ -38,10 +38,4 @@
 
 struct spdk_thread;
 
-/**
- * Force the current system thread to act as if executing the given SPDK thread.
- * This is only used for unit testing.
- */
-void spdk_set_thread(struct spdk_thread *thread);
-
 #endif /* SPDK_THREAD_INTERNAL_H_ */
