diff --git a/CONFIG b/CONFIG
index 89c34e90b..02ce04692 100644
--- a/CONFIG
+++ b/CONFIG
@@ -256,3 +256,6 @@ CONFIG_COPY_FILE_RANGE=n
 
 # liblz4 is available
 CONFIG_HAVE_LZ4=n
+
+# aio_rw_flags are enabled
+CONFIG_HAVE_AIO_RW_FLAGS=n
diff --git a/configure b/configure
index 26c9b0f4d..d8daedc37 100755
--- a/configure
+++ b/configure
@@ -860,6 +860,22 @@ if [[ $sys_name != "Linux" ]]; then
 	fi
 fi
 
+if echo -e '#include <linux/fs.h>\n' \
+        '#include <libaio.h>\n' \
+        '#include <stddef.h>\n' \
+        '#ifndef RWF_NOWAIT\n' \
+        '#error "No RWF_NOWAIT is defined"\n' \
+        '#endif\n' \
+        'int main(int argc, char **argv) {\n' \
+        'return offsetof(struct iocb, aio_rw_flags);\n}\n' \
+        | "${BUILD_CMD[@]}" -c - ; then
+        echo HAVE_AIO_RW_FLAGS=YES
+        CONFIG[HAVE_AIO_RW_FLAGS]="y"
+else
+        echo HAVE_AIO_RW_FLAGS=NO
+        CONFIG[HAVE_AIO_RW_FLAGS]="n"
+fi
+
 if [ "${CONFIG[RDMA]}" = "y" ]; then
 	if [[ ! "${CONFIG[RDMA_PROV]}" == "verbs" ]] && [[ ! "${CONFIG[RDMA_PROV]}" == "mlx5_dv" ]]; then
 		echo "Invalid RDMA provider specified, must be \"verbs\" or \"mlx5_dv\""
diff --git a/module/bdev/aio/bdev_aio.c b/module/bdev/aio/bdev_aio.c
index b51d6c83a..01914fb9d 100644
--- a/module/bdev/aio/bdev_aio.c
+++ b/module/bdev/aio/bdev_aio.c
@@ -64,7 +64,7 @@ struct file_disk {
 	struct spdk_bdev	disk;
 	char			*filename;
 	int			fd;
-#ifdef RWF_NOWAIT
+#ifdef SPDK_CONFIG_HAVE_AIO_RW_FLAGS
 	bool			use_nowait;
 #endif
 	TAILQ_ENTRY(file_disk)  link;
@@ -116,7 +116,7 @@ bdev_aio_open(struct file_disk *disk)
 {
 	int fd;
 	int io_flag = disk->readonly ? O_RDONLY : O_RDWR;
-#ifdef RWF_NOWAIT
+#ifdef SPDK_CONFIG_HAVE_AIO_RW_FLAGS
 	struct stat st;
 #endif
 
@@ -134,7 +134,7 @@ bdev_aio_open(struct file_disk *disk)
 
 	disk->fd = fd;
 
-#ifdef RWF_NOWAIT
+#ifdef SPDK_CONFIG_HAVE_AIO_RW_FLAGS
 	/* Some aio operations can block, for example if number outstanding
 	 * I/O exceeds number of block layer tags. But not all files can
 	 * support RWF_NOWAIT flag. So use RWF_NOWAIT on block devices only.
@@ -212,7 +212,7 @@ bdev_aio_submit_io(enum spdk_bdev_io_type type, struct file_disk *fdisk,
 		io_set_eventfd(iocb, aio_ch->group_ch->efd);
 	}
 	iocb->data = aio_task;
-#ifdef RWF_NOWAIT
+#ifdef SPDK_CONFIG_HAVE_AIO_RW_FLAGS
 	if (fdisk->use_nowait) {
 		iocb->aio_rw_flags = RWF_NOWAIT;
 	}
