diff --git a/configure.ac b/configure.ac
index 9c5e4739..8e4f134b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -259,6 +259,14 @@ AC_ARG_WITH([libunwind],
     AS_HELP_STRING([--with-libunwind=PATH],
         [specify path where libunwind include directory and lib directory can be found]))
 
+# --enable-stack-unwind
+AC_ARG_ENABLE([stack-unwind],
+[  --enable-stack-unwind@<:@=OPTS@:>@ enable stack unwinding, which is disabled by default.
+        yes|verbose  - enable stack unwinding.  Dump the raw stack information too
+        unwind-only  - enable stack unwinding.  Do not dump the raw stack information
+        no|none      - disable stack unwinding
+],,[enable_stack_unwind=no])
+
 # --with-papi
 AC_ARG_WITH([papi],
     AS_HELP_STRING([--with-papi=PATH],
